ffi/src/table_changes.rs

```diff
@@ -0,0 +1,798 @@
+    allocate_fn: AllocateStringFn,
+) -> NullableCvoid {
+    let table_changes_scan = unsafe { table_changes_scan.as_ref() };
+    let table_root = table_changes_scan.table_root().to_string();
+    allocate_fn(kernel_string_slice!(table_root))
+}
+
+/// Get the logical schema of the specified table changes scan.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle.
+#[no_mangle]
+pub unsafe extern "C" fn table_changes_scan_logical_schema(
+    table_changes_scan: Handle<SharedTableChangesScan>,
+) -> Handle<SharedSchema> {
+    let table_changes_scan = unsafe { table_changes_scan.as_ref() };
+    table_changes_scan.logical_schema().clone().into()
+}
+
+/// Get the physical schema of the specified table changes scan.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle.
+#[no_mangle]
+pub unsafe extern "C" fn table_changes_scan_physical_schema(
+    table_changes_scan: Handle<SharedTableChangesScan>,
+) -> Handle<SharedSchema> {
+    let table_changes_scan = unsafe { table_changes_scan.as_ref() };
+    table_changes_scan.physical_schema().clone().into()
+}
+
+pub struct ScanTableChangesIterator {
+    data: Mutex<Box<dyn Iterator<Item = DeltaResult<ScanResult>> + Send>>,
+    engine: Arc<dyn ExternEngine>,
+}
+
+#[handle_descriptor(target=ScanTableChangesIterator, mutable=false, sized=true)]
+pub struct SharedScanTableChangesIterator;
+
+impl Drop for ScanTableChangesIterator {
+    fn drop(&mut self) {
+        debug!("dropping ScanTableChangesIterator");
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn table_changes_scan_execute(
+    table_changes_scan: Handle<SharedTableChangesScan>,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<SharedScanTableChangesIterator>> {
+    let table_changes_scan = unsafe { table_changes_scan.as_ref() };
+    let engine = unsafe { engine.clone_as_arc() };
+    table_changes_scan_execute_impl(table_changes_scan, engine.clone())
+        .into_extern_result(&engine.as_ref())
+}
+
+fn table_changes_scan_execute_impl(
+    table_changes_scan: &TableChangesScan,
+    engine: Arc<dyn ExternEngine>,
+) -> DeltaResult<Handle<SharedScanTableChangesIterator>> {
+    let table_changes_iter = table_changes_scan.execute(engine.engine().clone())?;
+    let data = ScanTableChangesIterator {
+        data: Mutex::new(Box::new(table_changes_iter)),
+        engine: engine.clone(),
+    };
+    Ok(Arc::new(data).into())
+}
+
+/// # Safety
+///
+/// Drops table changes iterator.
+/// Caller is responsible for (at most once) passing a valid pointer returned by a call to
+/// [`scan_table_changes_execute`].
+#[no_mangle]
+pub unsafe extern "C" fn free_scan_table_changes_iter(
+    data: Handle<SharedScanTableChangesIterator>,
+) {
+    data.drop_handle();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn scan_table_changes_next(
+    data: Handle<SharedScanTableChangesIterator>,
+) -> ExternResult<ArrowFFIData> {
```

Comment:
We're checking that the iteration has ended by checking that the ArrowFFIData  is empty. I think we'll eventually want to change this too

Reply:
likely the FFI OptionalValue once that's merged. @samansmink is working on that, so we can do this as a followup.